petastorm/reader_impl/shuffling_buffer.py

```diff
@@ -0,0 +1,168 @@
+    def __init__(self):
+        self.store = []
+
+    def add_many(self, items):
+        self.store.extend(items)
+
+    def retrieve(self):
+        return self.store.pop(0)
+
+    def can_retrieve(self):
+        return len(self.store) > 0
+
+    def can_add(self):
+        return True
+
+    @property
+    def size(self):
+        return len(self.store)
+
+    def finish(self):
+        pass
+
+class RandomShufflingBuffer(ShufflingBufferBase):
+    """
+    A random shuffling buffer implementation. Items can be added to the buffer and retrieved in a random order.
+    """
+    def __init__(self, shuffling_buffer_capacity, min_after_retrieve, extra_capacity=1000):
+        """Initializes a new ShufflingBuffer instance.
+
+        Items may be retrieved from the buffer once `min_after_dequeue` items were added to the queue
+        (indicated by `can_retrieve`).
+
+        Items may be added to the buffer as long as the amount of items in the buffer does not exceed
+        `shuffling_queue_capacity`.
+
+        The amount of items in the buffer may grow above `shuffling_buffer_capacity` (due to a call to `add_many`
+        with a list of items), but must remain under `extra_queue_capacity`.
+
+        :param shuffling_buffer_capacity: Items may be added to the buffer as long as the amount of items in the
+        buffer does not exceed the value of `shuffling_queue_capacity`.
+        :param min_after_retrieve: The amount of items in the buffer may grow above `shuffling_buffer_capacity`
+        (due to a call to `add_many` with a list of items), but must remain under `extra_queue_capacity`
+        """
+        self._extra_capacity = extra_capacity
+        # Preallocate the shuffling buffer.
+        self._items = [None] * (shuffling_buffer_capacity + self._extra_capacity)
+        self._shuffling_queue_capacity = shuffling_buffer_capacity
+        self._min_after_dequeue = min_after_retrieve
+        self._size = 0
+        self._done_adding = False
+
+    def add_many(self, items):
+        if self._done_adding:
+            raise RuntimeError('Can not call add_many after done_adding() was called.')
+
+        if not self.can_add():
+            raise RuntimeError('Can not enqueue. Check the return value of "can_enqueue()" to check if more '
+                               'items can be added.')
+
+        # We leave self._extra_capacity slack to make sure we don't reallocate self._items array
+        expected_size = self.size + len(items)
+        maximal_capacity = self._shuffling_queue_capacity + self._extra_capacity
+        if expected_size > maximal_capacity:
+            raise RuntimeError('Attempt to enqueue more elements that the capacity allows. '
+                               'Current size: {}, new size {}, maximum allowed: {}'.format(self.size, expected_size,
+                                                                                           maximal_capacity))
+        self._items[self._size:len(items)] = items
```

Comment:
Shouldnt this be `self._items[self._size:self._size + len(items)] = items`

Reply:
Checked how come the tests don't catch this. To my surprise, the second parameter does not matter! It will just use the starting point. Updated the code with your proposal since it definitely shows the intent better.

Reply:
Looking at it, I think the idiomatic way to do it would be `self._items[self._size:] = ` though I think that either way is really fine.